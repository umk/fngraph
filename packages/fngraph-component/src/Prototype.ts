import { DataRecord, DataValue, Declaration, Primitive } from '@fngraph/data'

type Prototype<V = unknown> = DataValue extends V
  ? unknown
  :
      | Declaration
      | (V extends DataValue ? Matcher<V> : never)
      | (V extends Primitive ? V : never)
      | (V extends Array<infer I> ? Prototype<I> : never)
      | (V extends Record<string, unknown> ? { [K in keyof V]?: Prototype<V[K]> } : never)

/**
 * A matcher is a function, which takes a value and returns a collection of
 * records, where each record contains bindings to declarations for each
 * match in the value.
 * @param record A parent record that contains existing bindings
 * @param value The value to check against the matcher
 * @returns
 * A collection of records, where each record contains bindings to
 * declarations for each match in the value.
 */
export type Matcher<V extends DataValue> = (record: DataRecord, value: V) => Generator<DataRecord>

/**
 * A predicate is a function, which takes a value and returns
 * a boolean value, which indicates whether the value matches
 * the predicate.
 * @param value The value to check
 * @returns A boolean value, which indicates whether the value matches the predicate
 */
export type MatcherPredicate = (value: DataValue) => boolean

/**
 * Creates a matcher, which iterates values based on matches
 * against the predicate.
 * @param predicate The predicate, which is used to match values
 * @returns A matcher, which iterates values based on matches
 */
export function createPredicateMatcher<V extends DataValue>(
  predicate: MatcherPredicate,
): Matcher<V> {
  const matcher: Matcher<V> = function* (record, value) {
    if (Array.isArray(value)) {
      for (const item of value) {
        yield* matcher(record, item)
      }
    } else if (predicate(value)) {
      yield record
    }
  }
  return matcher
}

/**
 * A matcher, which considers any value to be a match.
 */
export const anyValue: Matcher<DataValue> = function* (record: DataRecord): Generator<DataRecord> {
  yield record
}

export default Prototype
