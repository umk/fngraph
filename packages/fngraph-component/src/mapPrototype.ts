import { Context, DataRecord, DataValue, Declaration } from '@fngraph/data'
import { OutgoingMapper } from '@fngraph/generator'

import Prototype from './Prototype'

/**
 * A function that matches provided statement value against
 * the prototype.
 * @param prototype A prototype that partially corresponds to the structure of matching value
 * @param value A value to match against the prototype
 * @param context The context object, which maps declarations to resolved values
 * @returns
 * A collection of contexts, where each context contains bindings
 * to declarations for each match in the value.
 */
function* getPrototypeMatches(
  prototype: Prototype,
  value: DataValue,
  context: Context = {} as Context,
): Generator<Context> {
  if (Array.isArray(value)) {
    // Prototypes may be defined against array value items,
    // but not against arrays themselves. The matcher just runs
    // a prototype comparison against each item of the array
    // value and returns records for matches.
    for (const item of value) {
      yield* getPrototypeMatches(prototype, item, { ...context })
    }
  } else {
    if (prototype instanceof Declaration) {
      // The prototype is a declaration, provided in an initial
      // query, so the value is either checked against the
      // declaration value, if it's present in the record, or
      // the value is set to the record.
      if (prototype.ensure(context, value)) {
        yield context
      }
    } else if (typeof prototype === 'function') {
      // The prototype is a custom matcher, which checks the
      // corresponding value in the structure of prototype by itself.
      yield* prototype(context, value)
    } else if (typeof prototype === 'object' && prototype !== null) {
      // The object prototype declares the shape of the value
      // and prototype for each of the properties. Even if the
      // prototype doesn't declare a single property, the value
      // is checked to be an object.
      if (typeof value === 'object' && value !== null) {
        yield* Object.entries(prototype).reduce(function* (prev, [key, prototype]) {
          if (key in value) {
            for (const context of prev) {
              yield* getPrototypeMatches(prototype, value[key], context)
            }
          }
        }, getThisContext(context))
      }
    } else if (prototype === value) {
      yield* getThisContext(context)
    }
  }
}

/**
 * Creates a generator that interates through a collection of
 * exactly one context.
 * @param context A context to iterate.
 */
function* getThisContext(context: Context): Generator<Context> {
  yield context
}

function mapPrototype<R extends DataRecord>(prototype: Prototype<R>): OutgoingMapper<R> {
  return function (record: R): Generator<Context> {
    return getPrototypeMatches(prototype, record)
  }
}

export default mapPrototype
