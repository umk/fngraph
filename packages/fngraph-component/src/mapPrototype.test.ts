import { faker } from '@faker-js/faker'
import { Declaration, DeclarationID } from '@fngraph/data'

import Prototype from './Prototype'
import mapPrototype from './mapPrototype'

describe('mapPrototype', () => {
  it('matches top level value', () => {
    const value = { property: faker.lorem.words(3) }
    const prototype: Prototype<typeof value> = {
      property: new Declaration('declaration' as DeclarationID),
    }
    const matches = mapPrototype<typeof value>(prototype)(value)
    expect(Array.from(matches)).toMatchSnapshot()
  })
  it('matches constants', () => {
    const valueA = { a: 'valueA', b: 100 }
    const valueB = { a: 'valueB', b: 200 }
    const prototype: Prototype<{
      a: string
      b: number
    }> = {
      a: new Declaration('declaration' as DeclarationID),
      b: 200,
    }
    const matches = [valueA, valueB].flatMap((s) => {
      const generator = mapPrototype<typeof s>(prototype)(s)
      return Array.from(generator)
    })
    expect(matches).toMatchSnapshot()
  })
  it('performs matching against array', () => {
    const value = {
      property: [faker.lorem.words(3), faker.lorem.words(3), faker.lorem.words(3)],
    }
    const prototype: Prototype<typeof value> = {
      property: new Declaration('declaration' as DeclarationID),
    }
    const matches = mapPrototype<typeof value>(prototype)(value)
    expect(Array.from(matches)).toMatchSnapshot()
  })
  it('performs partial structural comparison', () => {
    const value = {
      propertyA: {
        propertyA1: 'propertyA1',
        propertyA2: 'propertyA2',
        propertyA3: 'propertyA3',
      },
      propertyB: {
        propertyB1: 'propertyB1',
        propertyB2: {
          propertyB21: 'propertyB21',
          propertyB22: ['propertyB221', 'propertyB222', 'propertyB223'],
        },
      },
      propertyC: {
        propertyC1: 'propertyC1',
        propertyC2: 'propertyC2',
      },
    }
    const prototype: Prototype<typeof value> = {
      propertyA: {
        propertyA1: new Declaration('declarationA1' as DeclarationID),
        propertyA3: 'propertyA3',
      },
      propertyB: {
        propertyB1: 'propertyB1',
        propertyB2: {
          propertyB22: new Declaration('declarationB22' as DeclarationID),
        },
      },
      propertyC: {
        propertyC1: 'propertyC1',
      },
    }
    const matches = mapPrototype<typeof value>(prototype)(value)
    expect(Array.from(matches)).toMatchSnapshot()
  })
})
