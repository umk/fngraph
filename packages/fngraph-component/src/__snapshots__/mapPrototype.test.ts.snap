// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapPrototype matches constants 1`] = `
[
  {
    "declaration": "valueB",
  },
]
`;

exports[`mapPrototype matches top level value 1`] = `
[
  {
    "declaration": "civis pauper amiculum",
  },
]
`;

exports[`mapPrototype performs matching against array 1`] = `
[
  {
    "declaration": "civis pauper amiculum",
  },
  {
    "declaration": "amor ventosus cur",
  },
  {
    "declaration": "denuncio asporto rerum",
  },
]
`;

exports[`mapPrototype performs partial structural comparison 1`] = `
[
  {
    "declarationA1": "propertyA1",
    "declarationB22": "propertyB221",
  },
  {
    "declarationA1": "propertyA1",
    "declarationB22": "propertyB222",
  },
  {
    "declarationA1": "propertyA1",
    "declarationB22": "propertyB223",
  },
]
`;
